README.md
# Convolutional-Neural-Network-Circuit-Design
Final project of Integrated Circuit Design

# 1. RTL level design methodology
We maintain a table, which has a size of 4x66, to store the input data from the testbench. The width of the table is 66 because we can set all element of this table to zero at the reset stage of the program, and store the input data from table[1] to table[64]. By doing so, the zero-padding task is fulfilled automatically. After that, we calculate the result of convolution, then do the "ReLU" step, and store the result into the testbench. We request the input data from the testbench by using the address signal in the following order: 0, 64, 1, 65, 2, 66, ...... This can make sure that we can output the result to layer 0, layer 1, and layer 2 continuously. When all the data are generated successfully, the program will go to the idle state, waiting for the next image to process. The program will start to process the next image when the "ready" signal is triggered again.
Our RTL code can pass the simulation successfully.
  
# 2. Gate level simulation
Our CONV_syn.v can also pass the simulation under a clock cycle of 6ns. The following
are the area report and the simulation result.
   
# 3. Transistor level simulation
Our CONV_APR.v can pass the simulation under a clock cycle of 10.7ns. During
simulation, there will be lots of glitch suppression warnings. But that is acceptable during transistor level simulation. The following are the simulation result and the core area report.




![Simulation result and the simulation time needed.](time.png)
![Core area and total area of the chip needed.](area.png)

# 4. System Structure

Figure 1
![Figure1](system_structure1.png)

Figure 2
![Figure2](system_structure2.png)


INCA_libs/history
//
// File created by:  irun
// Do not modify this file

s1::(21Jun2019:16:17:24):( ncverilog testfixture.v CONV.v +access+r )
s2::(21Jun2019:16:18:09):( ncverilog testfixture.v CONV.v +access+r )
s3::(21Jun2019:16:19:15):( ncverilog testfixture.v CONV.v +access+r )
s4::(21Jun2019:16:19:42):( ncverilog testfixture.v CONV.v +access+r )
s5::(21Jun2019:16:21:09):( ncverilog testfixture.v CONV.v +access+r )
s6::(21Jun2019:16:22:29):( ncverilog testfixture.v CONV.v +access+r )
s7::(21Jun2019:16:28:44):( ncverilog testfixture.v CONV.v +access+r )
s8::(21Jun2019:16:31:31):( ncverilog testfixture.v CONV.v +access+r )
s9::(21Jun2019:16:34:26):( ncverilog testfixture.v CONV.v +access+r )
s10::(21Jun2019:16:35:25):( ncverilog testfixture.v CONV.v +access+r )
s11::(21Jun2019:16:37:59):( ncverilog testfixture.v CONV.v +access+r )
s12::(21Jun2019:16:39:17):( ncverilog testfixture.v CONV.v +access+r )
s13::(21Jun2019:16:39:27):( ncverilog testfixture.v CONV.v +access+r )
s14::(21Jun2019:16:44:52):( ncverilog testfixture.v CONV.v +access+r )
s15::(21Jun2019:16:48:53):( ncverilog testfixture.v CONV.v +access+r )
s16::(21Jun2019:16:52:18):( ncverilog testfixture.v CONV.v +access+r )
s17::(21Jun2019:16:52:42):( ncverilog testfixture.v CONV.v +access+r )
s18::(21Jun2019:16:53:05):( ncverilog testfixture.v CONV.v +access+r )
s19::(21Jun2019:16:55:37):( ncverilog testfixture.v CONV.v +access+r )
s20::(21Jun2019:16:56:31):( ncverilog testfixture.v CONV.v +access+r )
s21::(21Jun2019:16:58:28):( ncverilog testfixture.v CONV.v +access+r )
s22::(21Jun2019:16:59:12):( ncverilog testfixture.v CONV.v +access+r )
s23::(21Jun2019:16:59:38):( ncverilog testfixture.v CONV.v +access+r )
s24::(21Jun2019:17:00:45):( ncverilog testfixture.v CONV.v +access+r )
s25::(21Jun2019:17:01:13):( ncverilog testfixture.v CONV.v +access+r )
s26::(21Jun2019:17:02:41):( ncverilog testfixture.v CONV.v +access+r )
s27::(21Jun2019:17:06:03):( ncverilog testfixture.v CONV.v +access+r )
s28::(21Jun2019:17:08:31):( ncverilog testfixture.v CONV.v +access+r )
s29::(21Jun2019:17:11:55):( ncverilog testfixture.v CONV.v +access+r )
s30::(21Jun2019:17:12:13):( ncverilog testfixture.v CONV.v +access+r )
s31::(21Jun2019:17:12:25):( ncverilog testfixture.v CONV.v +access+r )
s32::(21Jun2019:17:12:40):( ncverilog testfixture.v CONV.v +access+r )
s33::(21Jun2019:17:13:51):( ncverilog testfixture.v CONV.v +access+r )
s34::(21Jun2019:17:20:34):( ncverilog testfixture.v CONV.v +access+r )
s35::(21Jun2019:17:20:49):( ncverilog testfixture.v CONV.v +access+r )
s36::(21Jun2019:17:21:13):( ncverilog testfixture.v CONV.v +access+r )
s37::(21Jun2019:17:30:44):( ncverilog testfixture.v CONV.v +access+r )
s38::(21Jun2019:17:36:09):( ncverilog testfixture.v CONV.v +access+r )
s39::(21Jun2019:17:39:49):( ncverilog testfixture.v CONV.v +access+r )
s40::(21Jun2019:17:43:34):( ncverilog testfixture.v CONV.v +access+r )
s41::(21Jun2019:17:44:59):( ncverilog testfixture.v CONV.v +access+r )
s42::(21Jun2019:17:45:10):( ncverilog testfixture.v CONV.v +access+r )
s43::(21Jun2019:17:46:26):( ncverilog testfixture.v CONV.v +access+r )
s44::(21Jun2019:17:49:18):( ncverilog testfixture.v CONV.v +access+r )
s45::(21Jun2019:17:51:19):( ncverilog testfixture.v CONV.v +access+r )
s46::(21Jun2019:17:54:45):( ncverilog testfixture.v CONV.v +access+r )
s47::(21Jun2019:17:55:30):( ncverilog testfixture.v CONV.v +access+r )
s48::(21Jun2019:18:46:28):( ncverilog testfixture.v CONV.v +access+r )
s49::(21Jun2019:18:47:21):( ncverilog testfixture.v CONV.v +access+r )
s50::(21Jun2019:18:53:13):( ncverilog testfixture.v CONV.v +access+r )
s51::(21Jun2019:18:53:37):( ncverilog testfixture.v CONV.v +access+r )
s52::(21Jun2019:18:54:57):( ncverilog testfixture.v CONV.v +access+r )
s53::(21Jun2019:18:55:06):( ncverilog testfixture.v CONV.v +access+r )
s54::(21Jun2019:18:55:29):( ncverilog testfixture.v CONV.v +access+r )
s55::(21Jun2019:19:02:07):( ncverilog testfixture.v CONV.v +access+r )
s56::(21Jun2019:19:02:51):( ncverilog testfixture.v CONV.v +access+r )
s57::(21Jun2019:19:05:44):( ncverilog testfixture.v CONV.v +access+r )
s58::(21Jun2019:19:09:55):( ncverilog testfixture.v CONV.v +access+r )
s59::(21Jun2019:19:20:54):( ncverilog testfixture.v CONV.v +access+r )
s60::(21Jun2019:19:21:11):( ncverilog testfixture.v CONV.v +access+r )
s61::(21Jun2019:19:22:51):( ncverilog testfixture.v CONV.v +access+r )
s62::(21Jun2019:19:24:59):( ncverilog testfixture.v CONV.v +access+r )
s63::(21Jun2019:19:27:54):( ncverilog testfixture.v CONV.v +access+r )
s64::(21Jun2019:19:29:21):( ncverilog testfixture.v CONV.v +access+r )
s65::(21Jun2019:19:29:43):( ncverilog testfixture.v CONV.v +access+r )
s66::(21Jun2019:19:31:16):( ncverilog testfixture.v CONV.v +access+r )
s67::(21Jun2019:19:32:24):( ncverilog testfixture.v CONV.v +access+r )
s68::(21Jun2019:19:35:46):( ncverilog testfixture.v CONV.v +access+r )
s69::(21Jun2019:19:40:10):( ncverilog testfixture.v CONV.v +access+r )
s70::(21Jun2019:19:40:28):( ncverilog testfixture.v CONV.v +access+r )
s71::(21Jun2019:19:41:08):( ncverilog testfixture.v CONV.v +access+r )
s72::(21Jun2019:19:42:14):( ncverilog testfixture.v CONV.v +access+r )
s73::(21Jun2019:19:42:45):( ncverilog testfixture.v CONV.v +access+r )
s74::(21Jun2019:19:45:34):( ncverilog testfixture.v CONV.v +access+r )
s75::(21Jun2019:19:48:46):( ncverilog testfixture.v CONV.v +access+r )
s76::(21Jun2019:19:49:11):( ncverilog testfixture.v CONV.v +access+r )
s77::(21Jun2019:19:56:46):( ncverilog testfixture.v CONV.v +access+r )
s78::(21Jun2019:19:58:27):( ncverilog testfixture.v CONV.v +access+r )
s79::(21Jun2019:19:59:06):( ncverilog testfixture.v CONV.v +access+r )
s80::(21Jun2019:20:00:07):( ncverilog testfixture.v CONV.v +access+r )
s81::(21Jun2019:20:00:24):( ncverilog testfixture.v CONV.v +access+r )
s82::(21Jun2019:20:02:24):( ncverilog testfixture.v CONV.v +access+r )
s83::(21Jun2019:20:17:06):( ncverilog testfixture.v -v tsmc13_neg.v CONV_syn.v +access+r +define+SDF )

CONV.v

`timescale 1ns/10ps

module  CONV(
        input wire clk,
	input wire reset,
	output reg busy,
	input wire ready,
        output reg [11:0] iaddr,
	input wire [19:0] idata,
	output reg cwr,
	output reg [11:0] caddr_wr,
	output reg [19:0] cdata_wr,
	output reg crd,
	output reg [11:0] caddr_rd,
	input wire [19:0] cdata_rd,
	output reg [2:0]  csel
	);


reg [6:0] indexx, indexy;
reg [1:0] sindex;

// Self-Defined registers
reg [3:0] state;
reg [19:0] buffer [3:0][65:0];
reg [35:0] string0 [3:0];
reg [35:0] string1 [3:0];
reg [35:0] l0temp1, l0temp2, l0temp3, l0temp4;
reg [6:0] center [1:0];
reg [11:0] l0addr, l1addr, l2addr;
reg [19:0] temp0, temp1;
reg write0;
wire [39:0] filter1 [8:0];
wire [39:0] filter2 [8:0];

integer i, j, k;

// Macro definition
parameter ini = 3'b000;
parameter layer0 = 3'b001;
parameter layer1 = 3'b010;
parameter layer2 = 3'b011;
parameter idle = 3'b100;
parameter layer0_calc = 3'b101;
parameter layer0_write = 3'b110;
parameter layer0_relu = 3'b111;
parameter layer0_calc2 = 4'b1000;
parameter layer0_calc3 = 4'b1001;
parameter idle2 = 4'b1010;
parameter idle3 = 4'b1011;
parameter bias1 = 40'h0000001310;      //Pixel 0: 7.446289e-02
parameter bias2 = 40'hFFFFFF7295;      //Pixel 1: -5.524139e-01

// Kernal weights
assign filter1[0] = 40'h000000A89E;      //Pixel 0: 6.586609e-01
assign filter2[0] = 40'hFFFFFFDB55;      //Pixel 1: -1.432343e-01
assign filter1[1] = 40'h00000092D5;      //Pixel 2: 5.735626e-01
assign filter2[1] = 40'h0000002992;      //Pixel 3: 1.623840e-01
assign filter1[2] = 40'h0000006D43;      //Pixel 4: 4.268036e-01
assign filter2[2] = 40'hFFFFFFC994;      //Pixel 5: -2.125854e-01
assign filter1[3] = 40'h0000001004;      //Pixel 6: 6.256104e-02
assign filter2[3] = 40'h00000050FD;      //Pixel 7: 3.163605e-01
assign filter1[4] = 40'hFFFFFF8F71;      //Pixel 8: -4.396820e-01
assign filter2[4] = 40'h0000002F20;      //Pixel 9: 1.840820e-01
assign filter1[5] = 40'hFFFFFF6E54;      //Pixel 10: -5.690308e-01
assign filter2[5] = 40'h000000202D;      //Pixel 11: 1.256866e-01
assign filter1[6] = 40'hFFFFFFA6D7;      //Pixel 12: -3.482819e-01
assign filter2[6] = 40'h0000003BD7;      //Pixel 13: 2.337494e-01
assign filter1[7] = 40'hFFFFFFC834;      //Pixel 14: -2.179565e-01
assign filter2[7] = 40'hFFFFFFD369;      //Pixel 15: -1.741791e-01
assign filter1[8] = 40'hFFFFFFAC19;      //Pixel 16: -3.277435e-01
assign filter2[8] = 40'h0000005E68;      //Pixel 17: 3.687744e-01

always@(posedge clk or posedge reset) begin
	if(reset) begin
		for(i = 0; i < 4; i = i + 1) begin
			for(j = 0; j < 66; j = j + 1) begin
				buffer[i][j] <= 20'b0;
			end
		end
		state <= idle;
		busy <= 0;
		indexx <= 7'd1;
		indexy <= 7'd1;
		center[0] <= 7'b1;
		center[1] <= 7'b1;
		sindex <= 2'b0;
		csel <= 3'b000;
		cwr <= 1'b0;
		crd <= 1'b0;
		l0addr <= 12'b0;
		l1addr <= 12'b0;
		l2addr <= 12'b0;
		write0 <= 1'b1;
	end
	else begin
		crd <= 1'b0;
		case (state)
			idle: begin
				state <= ini;
				iaddr <= 12'b0;
				busy <= 1;
			end
			idle2: begin
				busy <= 0;
				for(i = 0; i < 4; i = i + 1) begin
					for(j = 0; j < 66; j = j + 1) begin
						buffer[i][j] <= 20'b0;
					end
				end
				indexx <= 7'b1;
				indexy <= 7'b1;
				center[0] <= 7'b1;
				center[1] <= 7'b1;
				sindex <= 2'b0;
				csel <= 3'b000;
				cwr <= 1'b0;
				l0addr <= 12'b0;
				l1addr <= 12'b0;
				l2addr <= 12'b0;
				write0 <= 1'b1;
				state <= idle3;
			end
			idle3: begin
				if(ready == 1'b1) begin
					state <= idle;
				end
			end
			ini: begin
				buffer[indexx][indexy] <= idata;
				iaddr <= iaddr + 12'b1;
				indexy <= indexy + 7'b1;
				if(indexy == 7'd64) begin
					state <= layer0;
					indexx <= indexx + 1'b1;
					indexy <= 7'b1;
				end
			end
			layer0: begin
				buffer[indexx][indexy] <= idata;
				iaddr <= iaddr + 12'b1;
				indexy <= indexy + 7'b1;
				if(iaddr > 12'd4093 && indexy == 7'd64) begin
					state <= layer0_calc;
					for(k = 0; k < 66; k = k + 1) begin
						buffer[indexx + 1'b1][k] <= 20'b0;
					end
				end
				else if(indexx == 7'd3 && indexy == 7'd64) begin
					indexx <= 7'd2;
					indexy <= 7'b1;
					state <= layer0_calc;
				end
				else if(indexy == 7'd64) begin
					indexx <= indexx + 1'b1;
					indexy <= 7'b1;
				end
			end
			layer0_calc: begin
				cwr <= 1'b0;
				l0temp1 <= buffer[center[0]-1][center[1]-1]*filter1[0] + buffer[center[0]-1][center[1]]*filter1[1] + buffer[center[0]-1][center[1]+1]*filter1[2] + buffer[center[0]][center[1]-1]*filter1[3];
				l0temp2 <= buffer[center[0]-1][center[1]-1]*filter2[0] + buffer[center[0]-1][center[1]]*filter2[1] + buffer[center[0]-1][center[1]+1]*filter2[2] + buffer[center[0]][center[1]-1]*filter2[3];
				//string0[sindex] <= buffer[center[0]-1][center[1]-1]*filter1[0] + buffer[center[0]-1][center[1]]*filter1[1] + buffer[center[0]-1][center[1]+1]*filter1[2] + buffer[center[0]][center[1]-1]*filter1[3] + buffer[center[0]][center[1]]*filter1[4] + buffer[center[0]][center[1]+1]*filter1[5] + buffer[center[0]+1][center[1]-1]*filter1[6] + buffer[center[0]+1][center[1]]*filter1[7] + buffer[center[0]+1][center[1]+1]*filter1[8] + {bias1, 16'b0};
				//string1[sindex] <= buffer[center[0]-1][center[1]-1]*filter2[0] + buffer[center[0]-1][center[1]]*filter2[1] + buffer[center[0]-1][center[1]+1]*filter2[2] + buffer[center[0]][center[1]-1]*filter2[3] + buffer[center[0]][center[1]]*filter2[4] + buffer[center[0]][center[1]+1]*filter2[5] + buffer[center[0]+1][center[1]-1]*filter2[6] + buffer[center[0]+1][center[1]]*filter2[7] + buffer[center[0]+1][center[1]+1]*filter2[8] + {bias2, 16'b0};
				state <= layer0_calc2;
			end
			layer0_calc2: begin
				l0temp3 <= buffer[center[0]][center[1]]*filter1[4] + buffer[center[0]][center[1]+1]*filter1[5] + buffer[center[0]+1][center[1]-1]*filter1[6] + buffer[center[0]+1][center[1]]*filter1[7];
				l0temp4 <= buffer[center[0]][center[1]]*filter2[4] + buffer[center[0]][center[1]+1]*filter2[5] + buffer[center[0]+1][center[1]-1]*filter2[6] + buffer[center[0]+1][center[1]]*filter2[7];
				state <= layer0_calc3;
			end
			layer0_calc3: begin
				string0[sindex] <= l0temp1 + l0temp3 + buffer[center[0]+1][center[1]+1]*filter1[8] + {bias1, 16'b0};
				string1[sindex] <= l0temp2 + l0temp4 + buffer[center[0]+1][center[1]+1]*filter2[8] + {bias2, 16'b0};
				state <= layer0_relu;
			end
			layer0_relu: begin
				string0[sindex] <= string0[sindex][35] ? 36'b0 : string0[sindex];
				string1[sindex] <= string1[sindex][35] ? 36'b0 : string1[sindex];
				state <= layer0_write;
			end
			layer0_write: begin
				cwr <= 1;
				if(write0) begin
					csel <= 3'b001;
					caddr_wr <= l0addr;
					cdata_wr <= string0[sindex][15] ? string0[sindex][35:16] + 1'b1 : string0[sindex][35:16];
					write0 <= 1'b0;
				end
				else if(!write0) begin
					csel <= 3'b010;
					caddr_wr <= l0addr;
					cdata_wr <= string1[sindex][15] ? string1[sindex][35:16] + 1'b1 : string1[sindex][35:16];
					write0 <= 1'b1;
					if(sindex == 2'd3) begin
						sindex <= 2'b0;
						state <= layer1;
					end
					else begin
						sindex <= sindex + 2'b1;
						state <= layer0_calc;
						if(center[0] == 7'd1) begin
							center[0] <= 7'd2;
							l0addr <= l0addr + 12'd64;
						end
						else if(center[0] == 7'd2) begin
							center[0] <= 7'd1;
							center[1] <= center[1] + 7'd1;
							l0addr <= l0addr - 7'd63;
						end
					end
				end
			end
			layer1: begin
				if(write0) begin
					csel <= 3'b011;
					caddr_wr <= l1addr;
					write0 <= 1'b0;
					if((string0[0] >= string0[1]) && (string0[0] >= string0[2]) && (string0[0] >= string0[3])) begin
						cdata_wr <= string0[0][15] ? string0[0][35:16] + 1'b1 : string0[0][35:16];
						temp0 <= string0[0][15] ? string0[0][35:16] + 1'b1 : string0[0][35:16];
					end
					else if((string0[1] >= string0[0]) && (string0[1] >= string0[2]) && (string0[1] >= string0[3])) begin
						cdata_wr <= string0[1][15] ? string0[1][35:16] + 1'b1 : string0[1][35:16];
						temp0 <= string0[1][15] ? string0[1][35:16] + 1'b1 : string0[1][35:16];
					end
					else if((string0[2] >= string0[0]) && (string0[2] >= string0[1]) && (string0[2] >= string0[3])) begin
						cdata_wr <= string0[2][15] ? string0[2][35:16] + 1'b1 : string0[2][35:16];
						temp0 <= string0[2][15] ? string0[2][35:16] + 1'b1 : string0[2][35:16];
					end
					else begin
						cdata_wr <= string0[3][15] ? string0[3][35:16] + 1'b1 : string0[3][35:16];
						temp0 <= string0[3][15] ? string0[3][35:16] + 1'b1 : string0[3][35:16];
					end
				end
				else if(!write0) begin
					csel <= 3'b100;
					caddr_wr <= l1addr;
					write0 <= 1'b1;
					if((string1[0] >= string1[1]) && (string1[0] >= string1[2]) && (string1[0] >= string1[3])) begin
						cdata_wr <= string1[0][15] ? string1[0][35:16] + 1'b1 : string1[0][35:16];
						temp1 <= string1[0][15] ? string1[0][35:16] + 1'b1 : string1[0][35:16];
					end
					else if((string1[1] >= string1[0]) && (string1[1] >= string1[2]) && (string1[1] >= string1[3])) begin
						cdata_wr <= string1[1][15] ? string1[1][35:16] + 1'b1 : string1[1][35:16];
						temp1 <= string1[1][15] ? string1[1][35:16] + 1'b1 : string1[1][35:16];
					end
					else if((string1[2] >= string1[0]) && (string1[2] >= string1[1]) && (string1[2] >= string1[3])) begin
						cdata_wr <= string1[2][15] ? string1[2][35:16] + 1'b1 : string1[2][35:16];
						temp1 <= string1[2][15] ? string1[2][35:16] + 1'b1 : string1[2][35:16];
					end
					else begin
						cdata_wr <= string1[3][15] ? string1[3][35:16] + 1'b1 : string1[3][35:16];
						temp1 <= string1[3][15] ? string1[3][35:16] + 1'b1 : string1[3][35:16];
					end
					state <= layer2;
					l1addr <= l1addr + 1'b1;
				end
			end
			layer2: begin
				csel <= 3'b101;
				if(l2addr == 12'd2048) begin
					busy <= 0;
					state <= idle2;
				end
				else if(write0) begin
					caddr_wr <= l2addr;
					l2addr <= l2addr + 1'b1;
					write0 <= 1'b0;
					cdata_wr <= temp0;
				end
				else if(!write0) begin
					caddr_wr <= l2addr;
					l2addr <= l2addr + 1'b1;
					write0 <= 1'b1;
					cdata_wr <= temp1;
					if(l2addr == 2047) begin
						l2addr <= l2addr + 12'b1;
					end
					else if(center[0] == 7'd2 && center[1] == 7'd64) begin
                                                center[0] <= 7'd1;
                                                center[1] <= 7'd1;
                                                for(j = 0; j < 66; j = j + 1) begin
                                                        buffer[0][j] <= buffer[2][j];
                                                        buffer[1][j] <= buffer[3][j];
                                                end
                                                l0addr <= l0addr + 1'b1;
						state <= layer0;
                                        end
					else begin
						state <= layer0_calc;
						center[0] <= 7'd1;
						center[1] <= center[1] + 7'd1;
						l0addr <= l0addr - 7'd63;
					end
				end
			end
		endcase
	end
end
endmodule
